### DISCUSSION TITLE
Add support for pre-push tasks

### DISCUSSION THREAD

**Discussion by dshafik** [phpro/grumphp#943] [Ideas]
| Q               | A
| --------------- | ---
| Version         | GrumPHP 1.5.0
| Bug?            | no
| New feature?    | yes
| Question?       | no
| Documentation?  | no
| Related tickets | n/a

Currently GrumPHP supports being run as a `pre-commit` hook, however some tasks we want to run only `pre-push`. These would be time-consuming tasks such as `psalm` or running our test suite.

In our case, we want to run our test suite on bare metal in `pre-commit` (~20s), but in Docker in `pre-push` (~3m).

The way I see this functioning is adding new config options as siblings to `tasks`, `pre-commit` and `pre-push`, each either their own `tasks` sub-key. `pre-commit.tasks` would execute identically to `tasks` in it's functionality (and either could be used, for backwards compatibility), the second `pre-push.tasks` would run when the `git:pre-push` command is run.

You would also need to allow the other siblings of `tasks` (e.g. `git_hook_variables`, `process_timeout`, `fixer`, `parallel` etc.) as children of `pre-commit` and `pre-push`.

I understand this is a _large_ change, and I'm happy to try and spend some time on implementing it if there's interest?

**Comment by veewee** [comment]
Hello,

In times where github actions or other pipeline tools are very cheap, I personally don't see a lot of added value in waiting 3 minutes before you can push code to a remote repository. In my opinion, local git hooks should be able to finish in a reasonable time, blocking you from committing easy to spot issues. 

That being said, I do see some value in having pre-push instead of pre-commit in some cases.

We once tried adding pre-push support, but bounced into some harder problems.
Some context: https://github.com/phpro/grumphp/pull/258

The main problems here were:

- The architecture of GrumPHP wasn't really ready for this, it is now!
- We were facing some problems with detecting what files are being pushed. This is still a hard thing todo IIRC.


If we don't want to dive deep into these files problems, a more simpler solution would be to make the pre-push command do a `grumphp run` with a specific testsuite name. For example `pre-push`. That way, we will always get a full list of all files instead of only the once you are pushing. 

(We are already doing something similar for `pre-commit` and `commit-msg`: https://github.com/phpro/grumphp/blob/master/doc/testsuites.md#overriding-git-hook-test-suites )

That way we don't have to change any of the existing tasks and we could make the logic opt-in by not defining the testsuite.

There are some drawbacks on this approach, but I think it would make the basic feature easier to achieve. On top of my mind, it would primarily be: setting up the git hook files for this.

Let me know what you think and feel free to play around with the idea. We can discuss in more details in a PR if you like.

**Comment by JustusNBB** [comment]
Bump! How about just using `pre-push` instead of `pre-commit`?
Having a `pre-commit` hook can be super annoying when rebasing multiple commits.

**Reply by veewee** [reply]
Please read https://github.com/phpro/grumphp/discussions/943#discussioncomment-1583864

Nothing drastically changed in here since this comment: The discussion is open and you are free to play around with the idea :)

**Reply by JustusNBB** [reply]
If there was a PR I would indeed be able to take part in the discussion there, I'm not in position (time/priority) to make one though, sorry.

Before I retreat back to being an observer, here is my background:
I've found multiple workarounds in different projects of my company, moving or overwriting "faulty Grumphp hooks", most are about four years old (faulty for us because they need customisation).

Yesterday a colleague found this setting
```yml
grumphp:
    hooks_dir: './custom/hooks/folder'
```
But it does not copy the `pre-push` hook, so we still need to copy that with another script.
